mempool/mempool.go

```diff
@@ -0,0 +1,231 @@
+package mempool
+
+import (
+	"encoding/binary"
+	"errors"
+
+	"github.com/NethermindEth/juno/core"
+	"github.com/NethermindEth/juno/core/felt"
+	"github.com/NethermindEth/juno/db"
+	"github.com/NethermindEth/juno/encoder"
+)
+
+type ValidatorFunc func(*BroadcastedTransaction) error
+
+type BroadcastedTransaction struct {
+	Transaction   core.Transaction
+	DeclaredClass core.Class
+}
+
+const (
+	poolLengthKey = "poolLength"
+	headKey       = "headKey"
+	tailKey       = "tailKey"
+)
+
+// Pool stores the transactions in a linked list for its inherent FCFS behaviour
+type storageElem struct {
+	Txn      BroadcastedTransaction
+	NextHash *felt.Felt
+}
+
+type Pool struct {
+	db        db.DB
+	validator ValidatorFunc
+	txPushed  chan struct{}
+}
+
+func New(poolDB db.DB) *Pool {
+	return &Pool{
+		db:        poolDB,
+		validator: func(_ *BroadcastedTransaction) error { return nil },
+		txPushed:  make(chan struct{}, 1),
+	}
+}
```

Comment:
So running `go build -gcflags="-m" mempool.go` suggests that returning `*Pool` will cause an escape to the heap. Changing the return type to `Pool` doesn't seem to prevent an escape to the heap since `validator` is still allocated there.

Returning the pointer makes it cheaper to pass around Pool. But leaving it allocated to the heap means the garbage collector will have to check it every time, throughout the entire life time of the Juno program. Honestly, I'm not sure how big of a deal this is. If the main goal of this suggestion is to prevent `Pool` from being allocated to the heap, then we would have to decouple `validator` from `Pool`.

Reply:
In this case, where your struct has multiple, potentially complex members Id go with a pointer purely to avoid creating e .g undesired copies of things like mutexes now or in the future.

Reply:
I ran this locally on my PC and it says the function escapes to the heap, but I don't think the entire object does. Usually is good enough to trust the go compiler escape analysis, for example if you return by value an object big enough it will move it to the heap without any consideration, even if you meant to return it as a value type.

Regarding the object on the heap, if it is going to be there forever, then the runtime has optimization for those, where each object has kind of a lifetime. After it is checked multiple time, it starts being checked less recently until there is a point where it doesn't get checked at all until a few cycles have gone by.

@jeffrydegrande point is a valid one and I hadn't considered it. I've little experience with concurrent Go so I cannot provide much argument either in favor or against.

As a final point, an object which holds pointer to others is not uncommon. Think of go slices (`[]T`) where they are basically a triplet with (`len`, `cap`, `ptr`) where `len` is the len property, the `cap` how much memory they have reserved and `ptr` which point to the actual place in memory. The triplet (usually) lives on the stack while the elements of the array live on the heap.

At the end, I believe this specific case won't matter in the long term. I would suggest to return it as a value type (I believe this route would be novel to you) and see what happens. Worst case scenario, we switch it back. Nonetheless, @rianhughes choose what you think is best!